### ISSUE TITLE
opencv3 misconfigures cmake because brew incorrectly sets $PYTHONPATH

### ISSUE THREAD

**Issue by amckinlay** [Homebrew/homebrew-science#3329]
# Issue

This occurs when running `brew install opencv3 --with-openni2 --with-python3`, the issue does not occur without both `--with-openni2` and `--with-python3`.

Homebrew succeeds and correctly generates Python 3 bindings (or so I think); however, the following calls to `python3-config` and `python3` in `opencv3.rb` exit early.

``` ruby
if build.with? "python3"
  py3_config = `python3-config --configdir`.chomp
  py3_include = `python3 -c "import distutils.sysconfig as s; print(s.get_python_inc())"`.chomp
  py3_version = Language::Python.major_minor_version "python3"
  args << "-DPYTHON3_LIBRARY=#{py3_config}/libpython#{py3_version}.#{dylib}"
  args << "-DPYTHON3_INCLUDE_DIR=#{py3_include}"
end
```

Apparently, when compiling with `openni2` as a dependency, Homebrew sets the `PYTHONPATH` environment variable because `openni2.rb` `depends_on :python`. This `PYTHONPATH` somehow leaks to the build script`opencv2.rb` and we get the following output from `python3-config` and `python3` printed to the screen.

```
Your PYTHONPATH points to a site-packages dir for Python 2.x but you are running Python 3.x!
     PYTHONPATH is currently: "/usr/local/lib/python2.7/site-packages"
     You should `unset PYTHONPATH` to fix this.
Your PYTHONPATH points to a site-packages dir for Python 2.x but you are running Python 3.x!
     PYTHONPATH is currently: "/usr/local/lib/python2.7/site-packages"
     You should `unset PYTHONPATH` to fix this.
```

This output above is displayed to the terminal during the `brew install`, which can be confusing. In addition, `py3_config` and `py3_include` variables in `opencv3.rb` are not set because `python3-config` and `python3` exit. These variables eventually set CMake arguments in `opencv3.rb`, except in this case the argument parameters are wrong.

Here is the call to CMake:

```
==> cmake .. -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG -DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/opencv3/3.1.0_1 -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev -DBUILD_JASPER=OFF -DBUILD_JPEG=OFF -DBUILD_TIFF=OFF -DBUILD_OPENEXR=OFF -DBUILD_PNG=OFF -DBUILD_ZLIB=OFF -DCMAKE_OSX_DEPLOYMENT_TARGET= -DJPEG_INCLUDE_DIR=/usr/local/opt/jpeg/include -DJPEG_LIBRARY=/usr/local/opt/jpeg/lib/libjpeg.dylib -DBUILD_opencv_java=OFF -DBUILD_opencv_python2=ON -DBUILD_opencv_python3=ON -DWITH_1394=OFF -DWITH_EIGEN=ON -DWITH_FFMPEG=OFF -DWITH_GPHOTO2=OFF -DWITH_GSTREAMER=OFF -DWITH_JASPER=OFF -DWITH_OPENEXR=ON -DWITH_OPENGL=OFF -DWITH_QUICKTIME=OFF -DWITH_QT=OFF -DWITH_TBB=OFF -DWITH_VTK=OFF -DWITH_CUDA=OFF -DWITH_OPENNI2=ON -DPYTHON2_EXECUTABLE=/usr/bin/python -DPYTHON2_LIBRARY=/System/Library/Frameworks/Python.framework/Versions/2.7/lib/libpython2.7.dylib -DPYTHON2_INCLUDE_DIR=/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -DPYTHON3_LIBRARY=/libpython3.5.dylib -DPYTHON3_INCLUDE_DIR= -DENABLE_SSSE3=ON -DENABLE_SSE41=ON -DENABLE_SSE42=ON -DENABLE_AVX=ON
```

Note `-DPYTHON3_LIBRARY=/libpython3.5.dylib` and `-DPYTHON3_INCLUDE_DIR=` are incorrect.

Also, configuration from CMake implies Python 3 wrappers are not generated:

```
--   Python 3:
--     Interpreter:                 NO
--     Libraries:                   NO
--     numpy:                       NO (Python3 wrappers can not be generated)
--     packages path:               -
```

Note: for some reason I still appear to have the Python 3 wrapper (`/usr/local/opt/opencv3/lib/python3.5/site-packages/cv2.cpython-35m-darwin.so`) after installing, but I have not tested it. (edit: nvm)
# Gist-logs

https://gist.github.com/anonymous/c6606cf6521328ef0d59

As an aside, there are also some weird lines regarding Python 3 in the CMake log:

```
-- Could NOT find PythonInterp: Found unsuitable version "1.4", but required is at least "3.4" (found /usr/local/bin/python3)
-- Could NOT find PythonInterp: Found unsuitable version "1.4", but required is at least "3.2" (found /usr/local/bin/python3)
```

**Comment by amckinlay** [comment]
Seems like this issue is related to Homebrew/homebrew/pull/43724

**Comment by fish2000** [comment]
Relatedly, I am likewise seeing similar issues with this formula’s cmake args under a Python 2 regime – causing `FindPythonInterp.cmake` to prefer the system Python (`/usr/bin/python`) over a brewed version, and ultimately resulting in the build erroring out when the Clang invocation can’t find `<Python.h>` (?!) when trying to build the `cv3` Python 2 extension module.

**Comment by iMichka** [comment]
Fixed by #5424 (workaround), and fixed by  some other patches before that correctly set the different Python library paths. I'm closing this issue. Please open a new one if you still have trouble with this. Thanks.
